examples/matching_engine/src/lib.rs

```diff
@@ -0,0 +1,203 @@
+
+    fn from_custom_bytes(short_key: &[u8]) -> Result<Self, ViewError> {
+        let mut bytes = short_key.to_vec();
+        bytes.reverse();
+        let value = bcs::from_bytes(&bytes)?;
+        Ok(value)
+    }
+}
+
+impl Price {
+    pub fn revert(&self) -> Self {
+        Price {
+            price: u64::MAX - self.price,
+        }
+    }
+}
+
+pub fn product_price_amount(price: Price, count: Amount) -> Amount {
+    count.try_mul(price.price as u128).expect("product")
+}
+
+/// Transfer operation back to the owners
+#[derive(Clone)]
+pub struct Transfer {
+    /// Beneficiary of the transfer
+    pub owner: AccountOwner,
+    /// Amount being transferred
+    pub amount: Amount,
+    /// Index of the token being transferred (0 or 1)
+    pub token_idx: u32,
+}
+
+/// An identifier for a buy or sell order
+pub type OrderId = u64;
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub enum ModifyCount {
+    All,
+    Partial(Amount),
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub enum OrderNature {
+    /// A bid for buying token 1 and paying in token 0
+    Bid,
+    /// An ask for selling token 1, to be paid in token 0
+    Ask,
+}
+
+scalar!(OrderNature);
+
+#[derive(Clone, Debug, Serialize, Deserialize, SimpleObject)]
+pub struct KeyBook {
+    /// The corresponding price
+    pub price: Price,
+    /// The nature of the order
+    pub nature: OrderNature,
+    /// The owner used for checks
+    pub owner: AccountOwner,
+}
+
+#[derive(Default, Clone, Debug, Serialize, Deserialize, SimpleObject)]
+pub struct AccountInfo {
+    /// The list of orders
+    pub orders: BTreeSet<OrderId>,
+}
+
+/// The order entry in the order book
+#[derive(Clone, Debug, Deserialize, Serialize, SimpleObject)]
+pub struct OrderEntry {
+    /// The number of token1 being bought or sold
+    pub count: Amount,
+    /// The one who has created the order
+    pub owner: AccountOwner,
+    /// The order_id (needed for possible cancel or modification)
+    pub order_id: OrderId,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub enum Order {
+    /// Insertion of an order
+    Insert {
+        owner: AccountOwner,
+        count: Amount,
+        nature: OrderNature,
+        price: Price,
+    },
+    /// Cancelling of an order
+    Cancel {
+        owner: AccountOwner,
+        order_id: OrderId,
+    },
+    /// Modifying order (only decreasing is allowed)
+    Modify {
+        owner: AccountOwner,
+        order_id: OrderId,
+        cancel_count: Amount,
+    },
+}
+
+scalar!(Order);
+
+//
+// Now the types for the operation
+//
+
+#[derive(Clone, Copy, Debug, Deserialize, Serialize)]
+pub struct Parameters {
+    /// The token0 and token1 used for the matching engine
+    pub tokens: [ApplicationId<FungibleTokenAbi>; 2],
+}
+
+scalar!(Parameters);
+
+/// Operations that can be sent to the application.
+#[derive(Debug, Deserialize, Serialize)]
+pub enum Operation {
+    /// Some order being transmitted
+    Order { order: Order },
```

Comment:
We need a verb: `ExecuteOrder`, `PassOrder`, etc.
The comments should explain that the order is executed on the chain of the order book.

I wonder if in the future the app-id should support several order books instead of only one default on the creation chain.

Reply:
Ok, for `ExecuteOrder`.